File: app/src/main/java/com/jon/cotgenerator/ui/ServiceBoundActivity.java
package com.jon.cotgenerator.ui;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.jon.cotgenerator.service.CotService;
import com.jon.cotgenerator.utils.Notify;

import timber.log.Timber;

/* Simple base class shared between all app activities, which allows them all to receive state updates from the service. */
abstract class ServiceBoundActivity
        extends AppCompatActivity
        implements CotService.StateListener {

    protected CotService service;

    protected ServiceConnection serviceConnection = new ServiceConnection() {
        @Override public void onServiceConnected(ComponentName name, IBinder binder) {
            Timber.i("onServiceConnected");
            service = ((CotService.ServiceBinder) binder).getService();
            service.registerStateListener(ServiceBoundActivity.this);
            onStateChanged(service.getState(), null);
        }
        @Override public void onServiceDisconnected(ComponentName name) {
            Timber.i("onServiceDisconnected");
            service.unregisterStateListener(ServiceBoundActivity.this);
            service = null;
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        /* Start the service and bind to it */
        Intent intent = new Intent(this, CotService.class);
        startService(intent);
        bindService(intent, serviceConnection, BIND_AUTO_CREATE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (service != null) {
            service.registerStateListener(this);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (service != null) {
            service.unregisterStateListener(this);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (service != null) {
            service = null;
            if (serviceConnection != null) {
                unbindService(serviceConnection);
                serviceConnection = null;
            }
        }
    }

    protected View getRootView() {
        return findViewById(android.R.id.content);
    }

    @Override
    public void onStateChanged(CotService.State state, @Nullable Throwable throwable) {
        if (state == CotService.State.ERROR && throwable != null) {
            Notify.red(getRootView(), "Error: " + throwable.getMessage());
        }
    }
}

File: common/src/main/java/com/jon/common/ui/ServiceBoundActivity.kt
package com.jon.common.ui

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.jon.common.CotApplication
import com.jon.common.repositories.StatusRepository
import com.jon.common.service.CotService
import com.jon.common.service.CotService.ServiceBinder
import com.jon.common.service.ServiceState
import com.jon.common.utils.Notify
import com.jon.common.variants.Variant

abstract class ServiceBoundActivity : AppCompatActivity(),
        ServiceConnection {

    private val statusRepository = StatusRepository.getInstance()
    protected val viewModel: StateViewModel by viewModels()

    protected var service: CotService? = null

    protected fun startCotService() {
        /* Start the service and bind to it */
        val intent = Intent(this, CotService::class.java)
        startService(intent)
        bindService(intent, this, BIND_AUTO_CREATE)

        observeServiceStatus()
    }

    override fun onResume() {
        super.onResume()
        CotApplication.activityIsVisible = true
    }

    override fun onPause() {
        super.onPause()
        CotApplication.activityIsVisible = false
    }

    override fun onDestroy() {
        super.onDestroy()
        service?.let {
            service = null
            unbindService(this)
        }
    }

    override fun onServiceConnected(name: ComponentName, binder: IBinder) {
        service = (binder as ServiceBinder).service
        service?.initialiseFusedLocationClient()
    }

    override fun onServiceDisconnected(name: ComponentName) {
        service = null
    }

    protected fun getRootView(): View {
        return findViewById(android.R.id.content)
    }

    private fun observeServiceStatus() {
        statusRepository.getStatus().observe(this) {
            viewModel.currentState = it
            invalidateOptionsMenu()
            when (it) {
                ServiceState.RUNNING -> Notify.green(getRootView(), "Service is running")
                ServiceState.STOPPED -> Notify.blue(getRootView(), "Service is not running")
                ServiceState.ERROR -> Notify.red(getRootView(), "Error: ${ServiceState.errorMessage}")
                else -> throw IllegalArgumentException("Unknown service state '$it'")
            }
        }
    }
}